core/chains/evm/txmgr/finalizer.go

```diff
@@ -0,0 +1,275 @@
+	txStore finalizerTxStore,
+	client finalizerChainClient,
+	headTracker finalizerHeadTracker,
+) *evmFinalizer {
+	lggr = logger.Named(lggr, "Finalizer")
+	return &evmFinalizer{
+		lggr:         logger.Sugared(lggr),
+		chainId:      chainId,
+		rpcBatchSize: int(rpcBatchSize),
+		txStore:      txStore,
+		client:       client,
+		headTracker:  headTracker,
+		mb:           mailbox.NewSingle[*evmtypes.Head](),
+	}
+}
+
+// Start the finalizer
+func (f *evmFinalizer) Start(ctx context.Context) error {
+	return f.StartOnce("Finalizer", func() error {
+		f.lggr.Debugf("started Finalizer with RPC batch size limit: %d", f.rpcBatchSize)
+		f.stopCh = make(chan struct{})
+		f.wg.Add(1)
+		go f.runLoop()
+		return nil
+	})
+}
+
+// Close the finalizer
+func (f *evmFinalizer) Close() error {
+	return f.StopOnce("Finalizer", func() error {
+		f.lggr.Debug("closing Finalizer")
+		close(f.stopCh)
+		f.wg.Wait()
+		return nil
+	})
+}
+
+func (f *evmFinalizer) Name() string {
+	return f.lggr.Name()
+}
+
+func (f *evmFinalizer) HealthReport() map[string]error {
+	return map[string]error{f.Name(): f.Healthy()}
+}
+
+func (f *evmFinalizer) runLoop() {
+	defer f.wg.Done()
+	ctx, cancel := f.stopCh.NewCtx()
+	defer cancel()
+	for {
+		select {
+		case <-f.mb.Notify():
+			for {
+				if ctx.Err() != nil {
+					return
+				}
+				head, exists := f.mb.Retrieve()
+				if !exists {
+					break
+				}
+				if err := f.ProcessHead(ctx, head); err != nil {
+					f.lggr.Errorw("Error processing head", "err", err)
+					f.SvcErrBuffer.Append(err)
+					continue
+				}
+			}
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (f *evmFinalizer) DeliverLatestHead(head *evmtypes.Head) bool {
+	return f.mb.Deliver(head)
+}
+
+func (f *evmFinalizer) ProcessHead(ctx context.Context, head *evmtypes.Head) error {
+	ctx, cancel := context.WithTimeout(ctx, processHeadTimeout)
+	defer cancel()
+	_, latestFinalizedHead, err := f.headTracker.LatestAndFinalizedBlock(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to retrieve latest finalized head: %w", err)
+	}
+	return f.processFinalizedHead(ctx, latestFinalizedHead)
+}
+
+// Determines if any confirmed transactions can be marked as finalized by comparing their receipts against the latest finalized block
+func (f *evmFinalizer) processFinalizedHead(ctx context.Context, latestFinalizedHead *evmtypes.Head) error {
+	// Cannot determine finality without a finalized head for comparison
+	if latestFinalizedHead == nil || !latestFinalizedHead.IsValid() {
+		return fmt.Errorf("invalid latestFinalizedHead")
+	}
+	earliestBlockNumInChain := latestFinalizedHead.EarliestHeadInChain().BlockNumber()
+	f.lggr.Debugw("processing latest finalized head", "block num", latestFinalizedHead.BlockNumber(), "block hash", latestFinalizedHead.BlockHash(), "earliest block num in chain", earliestBlockNumInChain)
+
+	// Retrieve all confirmed transactions, loaded with attempts and receipts
+	unfinalizedTxs, err := f.txStore.FindConfirmedTxes(ctx, f.chainId)
```

Comment:
This can potentially be an extremely heavy query, that loads tons of state in-memory. Is there a better way to handle this?

Reply:
I'm noticing we're loading all the corresponding confirmed transactions in memory, and then we process them. I believe it would be better if these checks were inside a single query in the DB, for two reasons: first we would have to load less state and second the DB will optimize the filtering for us.